### PR TITLE
Pipes - Kimberley Zell - Hotel

### PR DESCRIPTION
**Pull request by kimpossible1**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | When making Blocks I had to decide if I wanted to make a separate class for Block Admin or handle it under the same Admin area as the regular Bookings.  I also had to decide if I wanted to make a separate class for Block Bookings.  I went with making a class for the Block itself, and then using my Booking class when making a block booking.  I kept the admin part with my Reservations class.  The functionality of each part of the Block helped me decide where to put it.|
| Describe a concept that you gained more clarity on as you worked on this assignment. | I think I better understand what initialize does and what can be put in initialize.  I also better understand what the attr reader vs accessor does. I had started using most att_reader from what we talked about in class, that it's important to not let things be write accessible if there isn't need for it, but I did not realize that would mean I also cannot access things that are not writeable in other parts of my code.  So after I understood that better, I started using attr_accessor more.  Overall I feel that I'm getting better with using classes and definitions.   |
| Describe a _nominal test_ that you wrote for this assignment. |  I wrote a test to make sure my program won't allow a room booking to go through if the room is already booked.  |
| Describe an _edge case test_ that you wrote for this assignment. | I wrote a test to make sure that my program won't allow a booking on a room block if the room is not reserved for the block.   |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? |I think I did okay.  I tend to sometimes write some code in with my pseudocode.  I'm pretty sure my code could be refactored to be more concise, and there must be a better way to organize.  The Reservations class became particularly long.  I perhaps could make more classes.   |

### CODE REVIEW

File: lib/Reservations.rb [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+    def initialize
+      @all_reservations = []
+      @rooms_collection = []
+      all_rooms
+      @blocks_collection = []
+    end
+
+    def all_rooms
+      @rooms_collection = []
+      n = 1
```
Comment (droberts-sea) [inline lib/Reservations.rb]: This method will recreate all the rooms from scratch every time it's called. This is a problem because any reference to a room from before won't register as equal to a room from after even if they have the same room number, because the object IDs will be different.

A better strategy might be to check if the list already exists at the beginning of this method, and if it does immediately return it.

File: lib/Reservations.rb [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    end
+
+    def all_reservations
+      return @all_reservations
+    end
+
+    def new_reservation(check_in, check_out, room_number = rand(1..20), room_rate = 200)
+      dates = Hotel::DateRange.new(check_in, check_out).dates
+      booking = Hotel::Booking.new(check_in, check_out, room_number, room_rate)
+      if check_availability?(dates, room_number) == false
```
Comment (droberts-sea) [inline lib/Reservations.rb]: All the places that you create a `DateRange`, it seems like you're only using it as a tool to get an array of dates. This indicates that a method might be a better choice here, one that takes a checkin and checkout date and returns an array of dates.

The other option would be to write an instance method, something like `DateRange#overlap?(other)`, that takes another `DateRange` and checks to see if they overlap. Then this complex bit of logic would be hidden away in `DateRange` instead of sitting here in `Reservations`.

As long as you've identified "manage a range of dates" as a responsibility, then it probably makes sense to do the later, and keep that responsibility out of the `Reservations` class.

This would also make it much easier to test!

File: lib/Booking.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        raise InvalidDateError.new("Check-in cannot be before today's date.")
+      end
+      unless @check_in < @check_out
+        raise InvalidDateError.new("Check-in cannot be after check-out.")
+      end
+    end
+
+    def total_cost
+      num_days = (@check_out - @check_in).to_i
+      total_cost = @room_rate * num_days
```
Comment (droberts-sea) [inline lib/Booking.rb]: Since you have this method defined here, you should not have an `attr_accessor` for `:total_cost` above (since this method will override the generated one).

File: lib/Booking.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  end
+
+  class Booking
+    attr_accessor :check_in, :check_out, :room_number, :total_cost, :dates
+    def initialize(check_in, check_out, room_number = rand(1..20), room_rate = 200)
+      @room_number = room_number
+      @check_in = Date.parse(check_in)
+      @check_out = Date.parse(check_out)
+      @date_range = DateRange.new(@check_in, @check_out)
+      @dates = @date_range.dates
```
Comment (droberts-sea) [inline lib/Booking.rb]: Instead of storing the check_in, check_out and dates here, it might be cleaner to only store the `DateRange` instance, and anyone who needs those attributes for a booking will have to go through `date_range`. That makes for a very clean separation of responsibilities: a `Booking` doesn't need to know what parts a `DateRange` has or how it works, it just knows that it has one.

File: lib/Booking.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      @room_rate = room_rate
+      #@dates = Hotel::DateRange.new(check_in, check_out)
+      # @num_days = (@check_out - @check_in).to_i
+      # @total_cost = (@room_rate * @num_days)
+      @discount = 0
+
+      validate_dates
+    end
+
+    def validate_dates
```
Comment (droberts-sea) [inline lib/Booking.rb]: I like that you run this validation right in the constructor, and that you've broken it out into a separate method. It might make even more sense as a check on the `DateRange` class, rather than having a `Booking` be responsible for this.

File: lib/Reservations.rb [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+        else
+          @all_reservations << booking
+          return booking
+      end
+    end
+
+    def list_rooms_available_by_date(date)
+      rooms_available = @rooms_collection
+      list_reservations_by_date(date).each do |booking|
+        rooms_available.each do |room|
```
Comment (droberts-sea) [inline lib/Reservations.rb]: I like this implementation, but this method ends up being very visually dense. Adding vertical whitespace (blank lines) and comments might help break it up into more digestible chunks.

File: lib/Reservations.rb [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+          if block.block_name == block_name
+            block.block_rooms_collection.each do |room|
+              if room.room_number == room_number
+              return true
+              end
+            end
+          end
+      end
+      raise ArgumentError.new("Room number #{room_number} not included in #{block_name} block.")
+      return false
```
Comment (droberts-sea) [inline lib/Reservations.rb]: It doesn't make any sense to `return false` after raising an `ArgumentError`. Because the `raise` stops the method execution, you'll never get to the `return`.

Of the two options, I think I prefer the `return` here. True or false are both viable answers to the question "is this room in this block" - I don't think there's any need for an error.

File: lib/Reservations.rb [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+            blocks_list << room
+          end
+        end
+      end
+      return blocks_list
+    end
+
+    def clear_reservations #Using this for testing purposes
+      @all_reservations = []
+    end
```
Comment (droberts-sea) [inline lib/Reservations.rb:209]: Since `@all_reservations` is an instance variable, I'm not sure this method is needed, even for testing. It would probably be cleaner to use a completely new instance for each test.

File: lib/Reservations.rb [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+        end
+      end
+    end
+
+    def check_block_room_available(block_name, room_number)
+      @blocks_collection.each do |block|
+        if block.block_name == block_name
+          block.booked.each do |booking|
+            if booking.room_number == room_number
+              raise ArgumentError.new("Unable to book. Room number #{room_number} has already been booked.")
```
Comment (droberts-sea) [inline lib/Reservations.rb]: This is an example of tight coupling between `Reservations` and `Block`. `Booking` reaches into the `Block` and looks at its data directly. A cleaner way to do things might be to write a method like `Block#has_room_available` and call it here.

Or you could redesign things so that when you reserve a room from a block you get the first available room (or an error if there are none). That removes a little bit of freedom from the user, but means that this check isn't needed at all.

File: lib/Reservations.rb [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+        end
+      end
+      return true
+    end
+
+    def list_reservations_by_date(date)
+      date = Date.parse(date)
+      list = []
+      @all_reservations.each do |reservation|
+        if date >= reservation.check_in && date < reservation.check_out  #check_out date excluded
```
Comment (droberts-sea) [inline lib/Reservations.rb]: This check is a great candidate for an instance method on `Booking`, or better yet, on `DateRange`.

File: specs/Reservations-specs.rb [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+      new_booking1.must_be_instance_of Hotel::Booking
+    end
+    it 'must raise an error if the check-in date is in the past' do
+
+      proc{new_booking = @new_hotel1.new_reservation("2017-06-01", "2017-09-21", 2)}.must_raise Hotel::InvalidDateError
+    end
+    it 'must raise an error if the check-in date is the same as the check-out date' do
+      proc{new_booking = @new_hotel1.new_reservation("2017-09-20", "2017-09-20"), 3}.must_raise Hotel::InvalidDateError
+    end
+    it 'must raise an error if the check_in date is after the check-out date' do
```
Comment (droberts-sea) [inline specs/Reservations-specs.rb]: Good coverage around error cases here.

File: specs/Reservations-specs.rb [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+    before do
+      @new_hotel.clear_reservations
+      @new_reservation1 = @new_hotel.new_reservation("2018-01-01", "2018-01-05", 1)
+    end
+    # after do
+    #   @new_hotel.clear_reservations
+    # end
+    it 'must must raise argument error if room is not available' do
+      proc{ new_reservation2 = @new_hotel.new_reservation("2018-01-01", "2018-01-04", 1) }.must_raise ArgumentError
+    end
```
Comment (droberts-sea) [inline specs/Reservations-specs.rb]: This is a good start, but I'd like to see more coverage around what date ranges will conflict with each other. Cases I'd be interested in:
- Not available:
  - Same dates
  - Overlaps in the front
  - Overlaps in the back
  - Completely contained
  - Completely containing
- Available:
  - Completely before
  - Completely after
  - Ends on the checkin date
  - Starts on the checkout date

File: specs/Reservations-specs.rb [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+    it 'must return an array' do
+      @new_hotel.list_rooms_available_by_date("2018-01-02").must_be_kind_of Array
+    end
+    it 'must return the correct number of rooms available' do
+      @new_hotel.list_rooms_available_by_date("2018-01-03").length.must_equal 17
+    end
+    it 'must return the correct rooms available' do
+      @new_hotel.list_rooms_available_by_date("2018-01-03")[0].room_number.must_equal 3
+    end
```
Comment (droberts-sea) [inline specs/Reservations-specs.rb]: Other cases I'd be interested in:
- What if there are no reservations? (it should return a list of all the rooms)
- Do reservations that don't overlap the date range affect the list? (they shouldn't)
  - Check the boundaries of what counts as an overlap. What about a reservation that ends on the checkin date, or starts on the checkout date?
- What if there are no vacancies? (it should return an empty array)
- What if your date range spans several small reservations? (all of the rooms from those reservations should be omitted from the list)

File: specs/Reservations-specs.rb [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+    it "must be an array" do
+      @new_hotel.list_reservations_by_date("2019-01-01").must_be_kind_of Array
+    end
+    it "must have the correct number of reservations for the date given" do
+      @new_hotel.list_reservations_by_date("2019-01-04").length.must_equal 3
+    end
+    it "must return the correct reservations" do
+      @new_hotel.list_reservations_by_date("2017-09-21")[0].room_number.must_equal @new_booking1.room_number
+    end
```
Comment (droberts-sea) [inline specs/Reservations-specs.rb]: You're missing some test cases here - they look very similar to the ones I would like to see for listing rooms by date.

File: specs/Reservations-specs.rb [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+      @new_block_reservation = @new_hotel_with_blocks1.new_reservation_in_block("2018-01-01", "2018-01-05", "Heritage", 6)
+      @new_block_reservation2 = @new_hotel_with_blocks1.new_reservation_in_block("2018-01-01", "2018-01-05", "Heritage", 7)
+    end
+    it 'must return true if there are rooms available in block to be booked' do
+      @new_block.has_rooms_available?.must_equal true
+    end
+    it 'must return false if there are NOT any rooms available in block to be booked' do
+      @new_block_reservation3 = @new_hotel_with_blocks1.new_reservation_in_block("2018-01-01", "2018-01-05", "Heritage", 8)
+      @new_block_reservation4 = @new_hotel_with_blocks1.new_reservation_in_block("2018-01-01", "2018-01-05", "Heritage", 9)
+      @new_block_reservation5 = @new_hotel_with_blocks1.new_reservation_in_block("2018-01-01", "2018-01-05", "Heritage", 5)
```
Comment (droberts-sea) [inline specs/Reservations-specs.rb]: Good failure case!

File: specs/Reservations-specs.rb [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+  end
+
+  describe 'list block rooms by date' do
+    before do
+      @new_hotel.clear_reservations
+      @new_block = @new_hotel.new_block("2018-01-01", "2018-01-10", "Heritage", 5)
+    end
+    it 'must list the correct number of rooms for a given date' do
+      @new_hotel.list_blocked_rooms_by_date("2018-01-02").length.must_equal 5
+    end
```
Comment (droberts-sea) [inline specs/Reservations-specs.rb]: Similar missing cases as for `Booking` above.

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| yes - good work
Answer comprehension questions	| yes
**Design** |
Each class is responsible for a single piece of the program | some - see my inline comments about `DateRange`
Classes are loosely coupled | mostly - I noticed a couple places where classes are hard to separate, but in general this looks good.
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | missing some cases around listing reservations and available rooms - see inline
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | missing many interesting cases around availability checking - see inline
**Wave 3** |
Create a block of rooms | yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage | missing some
**Additional Feedback** |
File Names | The ruby standard is that files should be named using snake_case - all lowercase, underscores between words. So instead of `lib/DateRange.rb` it would be `lib/date_range.rb`. Same for spec files, so the corresponding spec would be `specs/date_range_spec.rb`.
General Vibe | This is a good start, but it feels like there's a way to go still. The general pattern I notice is that your specific code within a method is usually solid, but design and test coverage could be improved. To me, this indicates you're having trouble taking a step back and imagining what your code looks like from the outside, thinking about what pieces ought to go where or what could go wrong. This important skill is difficult to master, but it's one you can grow with deliberate practice.
